### DISCUSSION TITLE
How to implement Widgets Provider in a vanilla UWP app?

### DISCUSSION THREAD

**Discussion by martinsuchan** [microsoft/WindowsAppSDK#3832] [Q&A] [answered]
Hi, I was thinking about adding Widgets support to my UWP apps, so I spent a day reading available documentation/guides and tried some samples, but it looks like it is not possible to implement Widgets Provider directly in my C#/UWP app, is that right?  

Widgets SDK right now requires the WinAppSDK dependency, and it is not possible to include the Microsoft.WindowsAppSDK nuget in my UWP app - I get error's that my app needs to target at least dotnet6.0. Also, I don’t plan to rewrite my UWP app to WinUI just to get Widgets support.  

I found this “guide” on how to include widgets in UWP app using some Desktop Bridge and COM server shenanigans, but it looks really complicated. https://nicksnettravels.builttoroam.com/uwp-widget/ 

So, my question is
- is there a way how to use Widgets from my vanilla UWP app directly, possibly in a similar way to windows.backgroundTasks?
- is there a plan to release nuget providing Widgets support to UWP apps?
- is there a plan to add native Widgets API into the UWP SDK? 
Thanks!

**Comment by hez2010** [comment]
I think you can refer to Widget Providers: https://learn.microsoft.com/en-us/windows/apps/develop/widgets/implement-widget-provider-cs
Basically you need to implement those COM interfaces and declare your implementation in the manifest to expose them.

**Reply by beervoley** [reply]
This won't work directly with UWP app since to expose those COM interface the application needs to be running in Medium IL :)

**Comment by beervoley** [comment, marked as answer]
Hi @martinsuchan, thanks for your interest in Widgets Platform :)

To answer your questions in order:
1) There's no way to integrate Widgets with the UWP app directly. Widget Providers require Medium IL (=Integrity Level) to run and access required COM functionality but UWP apps are being run in AppContainer which is Low IL by design.
So to implement and enable a WidgetProvider - you'll need an app that can run in Medium IL. The easiest way to get a WidgetProvider working with a UWP app is to use a [desktop extension](https://learn.microsoft.com/en-us/uwp/schemas/appxpackage/uapmanifestschema/element-desktop-extension). In essence, the idea is the same as you've already discovered in that guide: make a separate desktop app that will act as a `WidgetProvider` and communicate with it from your UWP app. Bidirectional communication is usually done via [AppServiceConnection](https://docs.microsoft.com/en-us/uwp/api/Windows.ApplicationModel.AppService.AppServiceConnection).
I know it may look rather complicated but really it's equivalent to adding `WidgetProvider` to your UWP application with which you will be communicating through an extra API Layer (=in this case it will be [AppServiceConnection](https://docs.microsoft.com/en-us/uwp/api/Windows.ApplicationModel.AppService.AppServiceConnection)).
There's a good article on how to make that `desktop extension` and communicate with it from your UWP app: https://stefanwick.com/2018/04/06/uwp-with-desktop-extension-part-1/ : )
2) There's currently no plan to directly enable UWP apps to be Widget Providers.
3) There's no plan to add Widgets API to UWP SDK.
